src/poll.rs

```diff
@@ -42,14 +48,133 @@ use std::time::Duration;
 /// poll.poll(None).unwrap();
 /// ```
 pub struct Poll {
+    // Platform specific IO selector
     selector: sys::Selector,
+
+    // Custom readiness queue
+    readiness_queue: ReadinessQueue,
+
+    // Pending events. TODO: remove this
     events: sys::Events,
 }
 
+#[derive(Clone, Debug)]
+pub struct Config {
+    // Timer tick size
+    timer_tick_dur: Duration,
+    // Size of the timer wheel
+    timer_wheel_size: usize,
+}
+
+pub struct Registration {
+    node: ReadyRef,
+    queue: ReadinessQueue,
+}
+
+#[derive(Clone)]
+struct ReadinessQueue {
+    inner: Arc<UnsafeCell<ReadinessQueueInner>>,
+}
+
+struct ReadinessQueueInner {
+    // All readiness nodes are owned by the `Poll` instance and live either in
+    // this linked list or in a `readiness_wheel` linked list.
+    head_all_nodes: Option<Box<ReadinessNode>>,
+
+    // linked list of nodes that are pending some processing
+    head_readiness: AtomicPtr<ReadinessNode>,
+
+    // Hashed timer wheel for delayed readiness notifications
+    readiness_wheel: Vec<AtomicPtr<ReadinessNode>>,
+
+    // Timer settings
+    timer_tick_ms: u64,
+
+    // Timer epoch
+    epoch: u64,
+
+    // Masks the target tick to get the slot in the wheel
+    mask: u64,
+}
+
+struct ReadyList {
+    head: ReadyRef,
+}
+
+struct ReadyRef {
+    ptr: *mut ReadinessNode,
+}
+
+struct ReadinessNode {
+    // ===== Fields only accessed by Poll =====
+    //
+    // Next node in ownership tracking queue
+    next_all_nodes: Option<Box<ReadinessNode>>,
+
+    // Previous node in the owned list
+    prev_all_nodes: ReadyRef,
+
+    // Current delay, represented in ticks. This also is used to find the
+    // "head" pointer of the ownership list that this node lives in.
+    delay: Option<Tick>,
+
+    // The Token used to register the `Evented` with `Poll`. This can change,
+    // but only by calling `Poll` functions, so there will be no concurrency.
+    token: Token,
+
+    // Both interest and opts can be mutated
+    interest: Cell<EventSet>,
+
+    // Poll opts
+    opts: Cell<PollOpt>,
```

Comment:
I'm personally pretty nervous about the direction this abstraction is headed in. As is it seems like it's incredibly difficult to get right in terms of preventing race conditions and other undefined behavior, and it's barely leveraging Rust types for managing this unsafety. Specifically this structure is conglomeration of:

* Raw unsafe pointers
* non-thread-safe `Cell` types
* threadsafe atomics

This, combined with the `fn inner_mut(&self) -> &mut ReadinessQueueInner` signature below, make this very difficult to determine whether it's actually safe or not.

I think this would benefit quite a bit from extracting some of the unsafe components here to standalone and isolated features with safe interfaces, or very very clear unsafe interfaces with documentation as to what contracts need to be upheld. For example `next_all_notes` and `prev_all_nodes` looks a lot like a doubly linked list? Does the standard library's `LinkedList` not suffice here because atomics are necessary? Some other reason?

Similarly if there's a set of fields that are only accessed by a `Poll` on each node, then this may benefit from a more concrete interface to indicate such. For example I would expect something like:

```rust
struct PollFields { ... }

struct ReadinessNode {
    poll_fields: UnsafeCell<PollFields>,
}

impl ReadinessNode {
    fn poll_fields<'a>(&'a self, _proof: &'a Poll) -> &'a PollFields { /* ... */ }
}
```

Basically something like that where the API itself is very difficult to get wrong if properly used. That should help anyone else working around here or perhaps any future refactorings as well.